spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/CassandraJoinRDD.scala

```diff
@@ -0,0 +1,152 @@
+      .filter(tableCol => columns.exists(joinCol => tableCol.columnName == joinCol.columnName))
+    if (chosenClusteringColumns.size != 0) {
+      val chosenClusteringIndexes = chosenClusteringColumns.flatMap(col => col.componentIndex)
+      val maxIndex = chosenClusteringIndexes.max
+      val requiredIndexes = 0 to maxIndex
+      val missingIndexes = requiredIndexes.toSet -- chosenClusteringIndexes.toSet
+      if (missingIndexes.size != 0) {
+        val maxCol = tableDef.clusteringColumns
+          .filter(cc => cc.componentIndex.get == maxIndex)
+          .map(_.columnName).mkString(", ")
+        val otherCols = tableDef.clusteringColumns
+          .filter(cc => missingIndexes.contains(cc.componentIndex.get))
+          .map(_.columnName).mkString(", ")
+        throw new IllegalArgumentException(s"Can't pushdown join on column ${maxCol} without also specifying [ ${otherCols} ]")
+      }
+    }
+    val missingPartitionKeys = partitionKeyColumnNames -- joinColumnNames
+    if (missingPartitionKeys.size != 0) {
+      throw new IllegalArgumentException(s"Can't join without the full partition key. Missing: [ ${missingPartitionKeys} ]")
+    }
+    columns.map(checkSingleColumn)
+  }
+
+  val rowWriter = implicitly[RowWriterFactory[O]].rowWriter(
+    tableDef,
+    joinColumnNames.map(_.columnName))
+
+  def on(joinColumns: ColumnSelector): CassandraJoinRDD[O, N] = {
+    new CassandraJoinRDD[O, N](prev, keyspaceName, tableName, connector, columnNames, joinColumns, where, readConf).asInstanceOf[this.type]
+  }
+
+  //We need to make sure we get selectedColumnNames before serialization so that our RowReader is
+  //built
+  private val singleKeyCqlQuery: (String) = {
+    require(tableDef.partitionKey.map(_.columnName).exists(
+      partitionKey => where.predicates.exists(_.contains(partitionKey))) == false,
+      "Partition keys are not allowed in the .where() clause of a Cassandra Join")
+    logDebug("Generating Single Key Query Prepared Statement String")
+    logDebug(s"SelectedColumns : $selectedColumnNames -- JoinColumnNames : $joinColumnNames")
+    val columns = selectedColumnNames.map(_.cql).mkString(", ")
+    val joinWhere = joinColumnNames.map(_.columnName).map(name => s"${quote(name)} = :$name")
+    val filter = (where.predicates ++ joinWhere).mkString(" AND ")
+    val quotedKeyspaceName = quote(keyspaceName)
+    val quotedTableName = quote(tableName)
+    val query = s"SELECT $columns FROM $quotedKeyspaceName.$quotedTableName WHERE $filter"
+    logDebug(query)
+    query
+  }
+
+  /**
+   * When computing a CassandraPartitionKeyRDD the data is selected via single CQL statements
+   * from the specified C* Keyspace and Table. This will be preformed on whatever data is
+   * avaliable in the previous RDD in the chain.
+   * @param split
+   * @param context
+   * @return
+   */
+  override def compute(split: Partition, context: TaskContext): Iterator[(O, N)] = {
+    logDebug(s"Query::: $singleKeyCqlQuery")
```

Comment:
It was already logged. Also, please keep logging messages consistent. Here it is prefixed with "Query", in the previous case it just logged naked query without any prefix.